lib/src/default_index_store.rs

```diff
@@ -730,10 +729,17 @@ impl Index for MutableIndexImpl {
 
     fn evaluate_revset<'index>(
         &'index self,
-        repo: &'index dyn Repo,
         expression: &RevsetExpression,
+        store: &Arc<Store>,
+        visible_heads: &[CommitId],
```

Comment:
Yes, I tried doing that, but it got complicated because of filtering. The problem is that filters implicitly depend on the visible heads, which means the revset engine needed to know the set of visible heads anyway. I think we should make the heads explicit in filters, and I started working on that, but then realized that it could be done later, so I sent only these changes for now.

By the way, I also experimented with replacing `IntersectionRevset` by `FilterRevset` by letting `ToPredicateFn` be able to terminate the iteration early. That worked mostly well, but it is measurably slower (~20% IIRC) on the `user(peff) & committer(gitster)` revset I added to the git.git benchmark suite.

Reply:
> filters implicitly depend on the visible heads

Right. Filters and unary negation have this problem. I'll take a look later. If we want to make `all()` include explicitly specified commits, `resolve_symbols()` would be a good place to.

> it is measurably slower (~20% IIRC) on the `user(peff) & committer(gitster)` revset

Interesting that filter predicates have a measurable impact. I also have unsent revset changes. Maybe it's time to revisit these as we now have a way to benchmark revset evaluation.

Reply:
FYI, here's that change to replace `IntersectionRevset` by `FilterRevset`: https://github.com/martinvonz/jj/tree/push-oquurlztknxo